packages/yew-router/src/router.rs

```diff
@@ -1,232 +1,111 @@
-    /// let render: Render<S> = Router::render(|switch: S| -> Html {
-    ///     match switch {
-    ///         S::Variant => html! {"Variant"},
-    ///     }
-    /// });
-    /// # }
-    /// ```
-    pub fn render<F: RenderFn<Router<SW, STATE>, SW> + 'static>(f: F) -> Render<SW, STATE> {
-        Render::new(f)
-    }
-
-    /// Wrap a redirect function so that it can be used by the Router.
-    pub fn redirect<F: RedirectFn<SW, STATE> + 'static>(f: F) -> Option<Redirect<SW, STATE>> {
-        Some(Redirect::new(f))
+    /// It is recommended that you use [`Router::render`] instead
+    pub fn new(value: impl Fn(R) -> Html + 'static) -> Self {
+        Self(Rc::new(value))
     }
 }
 
-/// Message for Router.
-#[derive(Debug, Clone)]
-pub enum Msg<STATE> {
-    /// Updates the route
-    UpdateRoute(Route<STATE>),
-}
-
-/// Render function that takes a switched route and converts it to HTML
-pub trait RenderFn<CTX: Component, SW>: Fn(SW) -> Html {}
-impl<T, CTX: Component, SW> RenderFn<CTX, SW> for T where T: Fn(SW) -> Html {}
-/// Owned Render function.
-#[derive(Clone)]
-pub struct Render<SW: Switch + Clone + 'static, STATE: RouterState = ()>(
-    pub(crate) Rc<dyn RenderFn<Router<SW, STATE>, SW>>,
-);
-impl<STATE: RouterState, SW: Switch + Clone> Render<SW, STATE> {
-    /// New render function
-    fn new<F: RenderFn<Router<SW, STATE>, SW> + 'static>(f: F) -> Self {
-        Render(Rc::new(f))
-    }
-}
-impl<STATE: RouterState, SW: Switch + Clone> Debug for Render<SW, STATE> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Render").finish()
+impl<T> Clone for RenderFn<T> {
+    fn clone(&self) -> Self {
+        Self(Rc::clone(&self.0))
     }
 }
 
-/// Redirection function that takes a route that didn't match any of the Switch variants,
-/// and converts it to a switch variant.
-pub trait RedirectFn<SW, STATE>: Fn(Route<STATE>) -> SW {}
-impl<T, SW, STATE> RedirectFn<SW, STATE> for T where T: Fn(Route<STATE>) -> SW {}
-/// Clonable Redirect function
-#[derive(Clone)]
-pub struct Redirect<SW: Switch + 'static, STATE: RouterState>(
-    pub(crate) Rc<dyn RedirectFn<SW, STATE>>,
-);
-impl<STATE: RouterState, SW: Switch + 'static> Redirect<SW, STATE> {
-    fn new<F: RedirectFn<SW, STATE> + 'static>(f: F) -> Self {
-        Redirect(Rc::new(f))
+impl<T> PartialEq for RenderFn<T> {
+    fn eq(&self, other: &Self) -> bool {
+        // https://github.com/rust-lang/rust-clippy/issues/6524
+        #[allow(clippy::vtable_address_comparisons)]
+        Rc::ptr_eq(&self.0, &other.0)
     }
 }
-impl<STATE: RouterState, SW: Switch> Debug for Redirect<SW, STATE> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Redirect").finish()
-    }
+
+/// Props for [`Router`]
+#[derive(Properties, Clone, PartialEq)]
+pub struct RouterProps<R: Clone> {
+    /// Callback which returns [`Html`] to be rendered for the current route.
+    pub render: RenderFn<R>,
 }
```

Comment:
`R` doesn't need to implement `Clone` here if you implement it manually.
```suggestion
pub struct RouterProps<R> {
    /// Callback which returns [`Html`] to be rendered for the current route.
    pub render: RenderFn<R>,
}

impl<R> Clone for RouterProps<R> {
    fn clone(&self) -> Self {
        Self { render: self.render.clone() }
    }
}
```